# Endpoint Chaining
---

The CDA provides a custom python tool for searching CDA data. [`Q`](usage/#q) (short for Query) offers several ways to search and filter data, and several input modes:

---
- **<a href="../../QuickStart/usage/#q">Q.()</a>** builds a query that can be used by `run()` or `count()`
- **<a href="../../QuickStart/usage/#qrun">Q.run()</a>** returns data for the specified search 
- **<a href="../../QuickStart/usage/#qcount">Q.count()</a>** returns summary information (counts) data that fit the specified search
- **<a href="../../QuickStart/usage/#columns">columns()</a>** returns entity field names
- **<a href="../../QuickStart/usage/#unique_terms">unique_terms()</a>** returns entity field contents

---
                                                                    
Before we do any work, we needs to import these functions cdapython.
We're also telling cdapython to report it's version so we can be sure we're using the one we mean to:

In [1]:
from cdapython import Q, columns, unique_terms, query
import cdapython
print(cdapython.__version__)

2022.6.22


In [ ]:
Q.set_default_project_dataset("broad-dsde-dev.cda_dev")
Q.set_host_url("https://cancerdata.dsde-dev.broadinstitute.org/")
Q.get_host_url()
Q.get_default_project_dataset()

## Endpoint Chaining

We're going to build on our <a href="../BasicSearch">previous basic search</a> to see what information exists about cancers that were first diagnosed in the brain. 


In [2]:
myquery = Q('ResearchSubject.primary_diagnosis_site = "brain"')


Previously we looked at subject, research_subject, specimen and file results seperately, but we can also combine these. 

Let's say what we're really interested in is finding analysis done on specimens, so we're looking for files that belong to specimens that match our search. If we search the files endpoint directly, we'll get back all files that meet our search criteria, regardless of whether they are specimens specficially. Instead, we can chain our query to the specimen endpoint and then to the files endpoint and get the combined result:

In [3]:
myqueryspecimenfiles =  myquery.specimen.file.run()
myqueryspecimenfiles

Getting results from database

Total execution time: 3535 ms



            QueryID: a5926a1e-ab3b-4000-8e2e-c94751782c84
            
            Offset: 0
            Count: 100
            Total Row Count: 405031
            More pages: True
            

We get back 405031 files that belong to specimens that meet our search criteria. As before, we can preview the results by using the `.to_dataframe()` function:

In [4]:
myqueryspecimenfiles.to_dataframe()

id  \
0   2c237a0e-58bf-4385-950e-3ef13b426a3e   
1   3e87a1bf-3752-4d7a-a064-1c992401ae6c   
2   3eecbeb9-ff63-4ae4-99f4-cf26bc78dcf1   
3   49306693-e7b4-45ce-afb2-93caf0d6dcae   
4   5845fe78-86bd-43a0-b365-b9673beada97   
..                                   ...   
95  76f4b413-aec3-4a8b-8109-24213551a9b8   
96  7f26a4ff-a560-4936-af95-854ee2dfab56   
97  870e2714-2a1e-477d-a9eb-2294de5112c0   
98  88d30ccb-5684-41fb-8e73-8877b9e5c683   
99  a82704c1-9be7-46c0-b0d5-1c2381a7d08a   

                                           identifier  \
0   [{'system': 'GDC', 'value': '2c237a0e-58bf-438...   
1   [{'system': 'GDC', 'value': '3e87a1bf-3752-4d7...   
2   [{'system': 'GDC', 'value': '3eecbeb9-ff63-4ae...   
3   [{'system': 'GDC', 'value': '49306693-e7b4-45c...   
4   [{'system': 'GDC', 'value': '5845fe78-86bd-43a...   
..                                                ...   
95  [{'system': 'GDC', 'value': '76f4b413-aec3-4a8...   
96  [{'system': 'GDC', 'value': '7f26a4ff-a560-493...   
97  [{'system': 'GDC', 'value': '870e2714-2a1e-477...   
98  [{'system': 'GDC', 'value': '88d30ccb-5684-41f...   
99  [{'system': 'GDC', 'value': 'a82704c1-9be7-46c...   

                                                label  \
0           TCGA-E1-A7YJ-01A-21-A44D-20_RPPA_data.tsv   
1           TCGA-QH-A6XA-01A-21-A44C-20_RPPA_data.tsv   
2           TCGA-S9-A6U5-01A-21-A44D-20_RPPA_data.tsv   
3           TCGA-HT-7603-01A-11-A29X-20_RPPA_data.tsv   
4           TCGA-HW-7489-01A-21-A29Y-20_RPPA_data.tsv   
..                                                ...   
95          TCGA-06-0190-01A-21-2214-20_RPPA_data.tsv   
96          TCGA-06-1087-01A-03-1900-20_RPPA_data.tsv   
97          TCGA-06-A6S0-01A-21-A44T-20_RPPA_data.tsv   
98                                seurat.analysis.tsv   
99  54bdb7e3-4a24-4bff-acb8-e51040063da9.genie.ali...   

                  data_category                          data_type  \
0            Proteome Profiling  Protein Expression Quantification   
1            Proteome Profiling  Protein Expression Quantification   
2            Proteome Profiling  Protein Expression Quantification   
3            Proteome Profiling  Protein Expression Quantification   
4            Proteome Profiling  Protein Expression Quantification   
..                          ...                                ...   
95           Proteome Profiling  Protein Expression Quantification   
96           Proteome Profiling  Protein Expression Quantification   
97           Proteome Profiling  Protein Expression Quantification   
98      Transcriptome Profiling               Single Cell Analysis   
99  Simple Nucleotide Variation  Masked Annotated Somatic Mutation   

   file_format associated_project  \
0          TSV           TCGA-LGG   
1          TSV           TCGA-LGG   
2          TSV           TCGA-LGG   
3          TSV           TCGA-LGG   
4          TSV           TCGA-LGG   
..         ...                ...   
95         TSV           TCGA-GBM   
96         TSV           TCGA-GBM   
97         TSV           TCGA-GBM   
98         TSV            CPTAC-3   
99         MAF          GENIE-MDA   

                                              drs_uri  byte_size  \
0   drs://dg.4DFC:2c237a0e-58bf-4385-950e-3ef13b42...      22030   
1   drs://dg.4DFC:3e87a1bf-3752-4d7a-a064-1c992401...      21962   
2   drs://dg.4DFC:3eecbeb9-ff63-4ae4-99f4-cf26bc78...      21988   
3   drs://dg.4DFC:49306693-e7b4-45ce-afb2-93caf0d6...      21989   
4   drs://dg.4DFC:5845fe78-86bd-43a0-b365-b9673bea...      21976   
..                                                ...        ...   
95  drs://dg.4DFC:76f4b413-aec3-4a8b-8109-24213551...      10511   
96  drs://dg.4DFC:7f26a4ff-a560-4936-af95-854ee2df...      24010   
97  drs://dg.4DFC:870e2714-2a1e-477d-a9eb-2294de51...      23970   
98  drs://dg.4DFC:88d30ccb-5684-41fb-8e73-8877b9e5...    5010388   
99  drs://dg.4DFC:a82704c1-9be7-46c0-b0d5-1c2381a7...       1051   

                            checksum

<div class="cdanote" style="background-color:#b3e5d5;color:black;padding:20px;">
    
<h3>Valid Endpoint Chains</h3>

Not all endpoints can be chained together. This is a restriction caused by the data itself. `diagnosis` and `treatment` information does not have files directly attached to it, instead these files are associated with the `researchsubject`, as such both "myquery.treatment.files.run()" and "myquery.diagnosis.files.run()" will fail, as there are no files to retrieve. Valid chains are:
    
<ul>
<li><b>myquery.subject.file.run:</b> This will return all the files that meet the query and that are directly tied to subject</li>
<li><b>myquery.researchsubject.file.run:</b>This will return all the files that meet the query and that are directly tied to researchsubject</li>
<li><b>myquery.specimen.file.run:</b> This will return all the files that meet the query and that are directly tied to specimen</li>
<li><b>myquery.subject.file.count.run:</b> This will return the count of files that meet the query and that are directly tied to subject</li>
<li><b>myquery.researchsubject.file.count.run:</b>This will return the count of files that meet the query and that are directly tied to researchsubject</li>
<li><b>myquery.specimen.file.count.run:</b> This will return the count of files that meet the query and that are directly tied to specimen</li>
</ul>
</div>